### PR TITLE
WIP: Changes to enable development environment to use containers

### PR DESCRIPTION
**Pull request by mpaluru**
This is still a WIP, hence not squashing the commits yet. Once the changes are accepted, I will recreate the PR with all the following requirements satisfied.

#### Pre-Submission Checklist

- [x] Your pull request targets the `staging` branch of FreeCodeCamp.
- [x] Branch starts with either `fix/`, `feature/`, or `translate/` (e.g. `fix/signin-issue`)
- [ ] You have only one commit (if not, [squash](http://forum.freecodecamp.com/t/how-to-squash-multiple-commits-into-one-with-git/13231) them into one commit).
- [ ] All new and existing tests pass the command `npm run test-challenges`. Use `git commit --amend` to amend any fixes.
#### Type of Change

- [ ] Small bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds new functionality)
- [ ] Breaking change (fix or feature that would change existing functionality)
- [ ] Add new translation (feature adding new translations)
#### Checklist:

- [x] Tested changes locally.
- [ ] Closes currently open issue (replace XXXX with an issue no): Closes #XXXX
#### Description

These changes are enable the use of containers for development environment.
#### TODO:
- Currently the image is pointed to mpaluru/fcc-dev. Need to move it to freecodecamp account
- Also need to setup a hook on docker hub to enable the rebuild of image for commits on the staging branch.

### CODE REVIEW

**Comment by BerkeleyTrue** [comment]
Is it possible to run this without docker?

File: gulpfile.js [hunk 1]
```diff
@@ -201,7 +201,7 @@ var syncDepenedents = [
 
 gulp.task('sync', syncDepenedents, function() {
   sync.init(null, {
-    proxy: 'http://localhost:3000',
+    proxy: 'http://0.0.0.0:3000',
```
Comment (BerkeleyTrue) [inline gulpfile.js]: Why does docker require 0.0.0.0 as the IP instead of 127.0.01?

File: Dockerfile.dev [hunk 2]
```diff
@@ -0,0 +1,23 @@
+FROM node:6.6.0
```
Comment (BerkeleyTrue) [inline Dockerfile.dev]: Can we add this files to a /docker folder?

**Comment by leodotcloud** [comment]
Is it possible to run this without docker?
- Nope, we need to install docker to be able to run the containers. For linux users it's literally one short command to install docker `curl -sSL https://get.docker.com | sh -`. For OS X users, it's a .dmg install and for Windows I assume it's a single exe file installation.

Why does docker require 0.0.0.0 as the IP instead of 127.0.01?
- In the host mode, 127.0.0.1 refers to the same host, but when running inside a container it's the localhost of the container but not the host. So to be able to open the communication from the host to the container I had to change it to 0.0.0.0

Can we add these files to a /docker folder?
- `Dockerfile` can copy files into the image only from the current folder, so if we move to docker folder, we would have to refer to the files using `../` and it's not supported. The other option could be moving all the current source files inside `FreeCodeCamp/src` and leave `FreeCodeCamp/Dockerfile` at level to keep the structure clean. If you look at other projects, `Dockerfile` is usually at the top level. Another common practice if we want to keep the `Dockerfile` inside the docker folder would be to generate one bundle file for client and server, using the build script copy them to `docker/build` folder and then in the `Dockerfile` reference them as `./build/client_bundle.js` or `./build/server_bundle.js`.

**Comment by mpaluru** [comment]
Is it possible to run this without docker?
- Nope, we need to install docker to be able to run the containers. For linux users it's literally one short command to install docker `curl -sSL https://get.docker.com | sh -`. For OS X users, it's a .dmg install and for Windows I assume it's a single exe file installation.

Why does docker require 0.0.0.0 as the IP instead of 127.0.01?
- In the host mode, 127.0.0.1 refers to the same host, but when running inside a container it's the localhost of the container but not the host. So to be able to open the communication from the host to the container I had to change it to 0.0.0.0

Can we add these files to a /docker folder?
- `Dockerfile` can copy files into the image only from the current folder, so if we move to docker folder, we would have to refer to the files using `../` and it's not supported. The other option could be moving all the current source files inside `FreeCodeCamp/src` and leave `FreeCodeCamp/Dockerfile` at level to keep the structure clean. If you look at other projects, `Dockerfile` is usually at the top level. Another common practice if we want to keep the `Dockerfile` inside the docker folder would be to generate one bundle file for client and server, using the build script copy them to `docker/build` folder and then in the `Dockerfile` reference them as `./build/client_bundle.js` or `./build/server_bundle.js`.

**Comment by BerkeleyTrue** [comment]
> Nope, we need to install docker to be able to run the containers.

So the only way to run FCC would be in a container? 

> can copy files into the image only from the current folder, 

Then can this be a separate repo instead of in this main repo?

**Comment by leodotcloud** [comment]
> So the only way to run FCC would be in a container?
The usual way of running locally will still be available. This is an extra option for people who do not want to install all the dependencies locally and for those who want to use containers.

> Then can this be a separate repo instead of in this main repo?
Yes it can be but separating it out into a separate repo is adding an extra step/complexity.
- Imagine if someone wants to rebuild the container, they would have to now clone two different repos and that starting the container with the path of the previous one, it's a little too complicated and not scriptable in a generic way.
- The general practice across the industry is to keep the Dockerfile together with the source code.

**Comment by mpaluru** [comment]
> So the only way to run FCC would be in a container?

The usual way of running locally will still be available. This is an extra option for people who do not want to install all the dependencies locally and for those who want to use containers.

> Then can this be a separate repo instead of in this main repo?

Yes it can be but separating it out adds an extra step/complexity.
- Imagine if someone wants to rebuild the container, they would have to now clone two different repos and then starting the container with the volume path of the fcc repo, it's a little too complicated and not scriptable in a generic way.
- The general practice across the industry is to keep the Dockerfile together with the source code in the same repo.

**Comment by BerkeleyTrue** [comment]
> everything was loading except the code editor. 

This makes it sound like the issue is not with the ip but the build dev tools. Do the rest of the static files load?

**Comment by mpaluru** [comment]
| This makes it sound like the issue is not with the ip but the build dev tools. Do the rest of the static files load?

Initially I thought the same but just changing the IP makes it work without any need for rebuilding.

**Comment by BerkeleyTrue** [comment]
> Initially I thought the same but just changing the IP makes it work without any need for rebuilding.

This is a possible solution, but not the cause. You must first identify the cause before you can form the proper response.

> Do the rest of the static files load?

Also, I've made significant changes to the dev server. Please rebase off the current staging

File: CONTRIBUTING.md [hunk 3]
```diff
@@ -6,6 +6,7 @@
 -   [Contribution Guidelines](#contribution-guidelines)
 -   [Prerequisites](#prerequisites)
 -   [Getting Started](#getting-started)
+-   [Development Environment using Containers] (#development-environment-using-containers)
```
Comment (atjonathan) [inline CONTRIBUTING.md]:
~~~`[Development Environment using Containers] (#development-environment-using-containers)`~~
`[Development Environment using Containers](#development-environment-using-containers)`

**Review by ghost — CHANGES_REQUESTED** [review]

**Comment by BerkeleyTrue** [comment]
@mpaluru ping

**Comment by mpaluru** [comment]
@BerkeleyTrue Have been busy, will try to work on this over the weekend.

**Comment by atjonathan** [comment]
@mpaluru, any update on this?

File: gulpfile.js (hunk 1 above)
Comment (raisedadead) [inline gulpfile.js]: I haven't used docker, but yeah on C9, you do have to setup 0.0.0.0 as your IP. Might I suggest we make this an ENV var and add a `require('dotenv').load();` to the top.

Of course with fallbacks to localhost ?

**Comment by BerkeleyTrue** [comment]
Closing as stale. Please re-open when ready.

Happy Coding

**Pull request closed without merging**